docs/RFCS/certificate_rotation.md

```diff
@@ -0,0 +1,293 @@
+
+If two certificates of the same type are present (eg: two CA certificates), report
+the timestamps for the certificate with the latest expiration date.
+
+With such metrics, we can now alert on expiring certificates, either with a fixed lifetime
+remaining or when a fraction of the lifetime has expired.
+
+It may be preferable to monitor certificate chains rather than individual certificates
+(see Known Issues below) and report validity dates for the latest valid chain.
+
+# Certificate creation and renewal
+
+## CA certificate
+
+Initial CA creation involves creating the certificate and private key.
+* `ca.cert`: the CA certificate, valid 5 years. Provided to all nodes and clients.
+* `ca.key`: the CA private key. Must be kept safe and **not** distributed to nodes and clients.
+
+CA renewal involves creating a new certificate using the same private key.
+We have two options:
+* add the new certificate to `ca.cert` and provide the expanded file to all nodes and clients.
+* create a new certificate (eg: `ca.new.cert`) and provide it to all nodes and clients.
+
+The key **should not** be regenerated on certificate renewal. If it is (human error or lost key),
+all client and node certificates will need to be regenerated and signed with the new key.
+
+## Node certificate
+
+The trusted machine holding the CA private key generates node certificates and private keys, then
+pushes them to the appropriate nodes.
+Keys are not kept by the trusted machine once deployed on the nodes.
+
+Upon renewal, filenames should differ allowing multiple sets of certificate/private key files
+to exist simultaneously.
+
+# Reloading certificate/key files
+
+The current method for providing certificates and private keys requires naming them specifically
+using flags or environment variables:
+* `--ca-cert`, `--ca-key` for CA certificate and key
+* `--cert`, `--key` for node or client certificate and key
+
+We can use a single file for CA certificates (see CA certificate renewal above) but we require
+multiple files for node and client certificates due to changing keys.
+
+We thus need to be able to specify a storage location containing an arbitrary number of
+certificates and keys.
+
+We propose a single directory containing all relevant certificates and private keys for
+the current node or client.
+
+## Directory
+
+The directory can be specified by command-line flag `--certs-dir` or environment variable
+`COCKROACH_CERTS_DIR`, with a default value of `cockroach-certs`.
+
+The permissions on the directory cannot be more permissive than `drwx------`.
+Directory ownership will not be checked (too much incompatibility across filesystems and
+architectures), relying instead on permissions to list the directory contents.
+
+## Files
+
+File requirements are as follows:
+* all files must be in the `COCKROACH_CERTS_DIR` directory, sub-directories are ignored
+* all files must be plain files, no sym-links allowed
```

Comment:
This restriction is unnecessary. Just say that  the realname of the file must have that of the containing directory as prefix.